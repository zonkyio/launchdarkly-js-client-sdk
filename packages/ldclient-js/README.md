# LaunchDarkly SDK - Main JavaScript Package

This file contains the `ldclient-js` package, which is the standard interface for the LaunchDarkly SDK for browser JavaScript. It is described in the [main README](../../README.md).

This package automatically imports `ldclient-js-common`, which provides much of the underlying implementation.
