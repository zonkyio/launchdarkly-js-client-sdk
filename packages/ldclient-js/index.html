<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge;chrome=1" />
    <script src="https://cdn.jsdelivr.net/npm/es6-promise@4/dist/es6-promise.auto.min.js"></script>
    <script src="https://unpkg.com/event-source-polyfill@0.0.12/src/eventsource.min.js"></script>
    <script src="./dist/ldclient.js"></script>
  </head>
  <body>
    <h1>Production Settings</h1>
    <ul>
      <li><code>client-side-flag</code>: <span class="prod-client-side-flag">initializing…</span></li>
      <li><code>another-client-side-flag</code>: <span class="prod-another-client-side-flag">initializing…</span></li>
    </ul>
    
    <button class="prod-switch-user">Change user</button>
    <button class="prod-change-hash">Change hash fragment</button>
    <button class="prod-change-url">Change url</button>

    <h1>Test Settings</h1>
    <ul>
      <li><code>client-side-flag</code>: <span class="test-client-side-flag">initializing…</span></li>
      <li><code>another-client-side-flag</code>: <span class="test-another-client-side-flag">initializing…</span></li>
    </ul>
    
    <button class="test-switch-user">Change user</button>
    <button class="test-change-hash">Change hash fragment</button>
    <button class="test-change-url">Change url</button>
    
    <script>
      var switched = false;
      var user = {key: 'foobar.com'};
      var clientProd = LDClient.initialize('569f514183f2164430000004', user);
      var clientTest = LDClient.initialize('569f514183f2164430000005', user, {
        bootstrap: {
          'client-side-flag': false,
          'another-client-side-flag': false
        },
        hash: 'c31e052b7b474e6e5f0c182144907e6fb6e55f20f5e7bc41b0b61624b16b5316',
        baseUrl: 'http://localhost',
        eventsUrl: 'http://localhost',
        streamUrl: 'http://localhost:5050',
        useReport: true
      });

      function main(env) {
        render(env, 'client-side-flag', false);
        render(env, 'another-client-side-flag', false);
        // setTimeout(main, 2000);
      }
      
      function render(env, key, defaultValue) {
        var value = clientProd.variation(key, defaultValue);
        var el = document.querySelector('.' + env + '-' + key);
        el.innerHTML = value.toString();
      }

      //production
      clientProd.on('ready', function() {
        console.log('client is ready');
        main('prod');
      });
      
      clientProd.on('change', function(settings) {
        console.log('flags changed:', settings);
        main('prod');
      });
      
      clientProd.on('change:another-client-side-flag', function(value, previous) {
        console.log('another-client-side-flag changed:', value, '(' + previous + ')');
        main('prod');
      });
    
      
      document.querySelector('.prod-switch-user').addEventListener('click', function(event) {
        var key = switched ? 'foobar.com' : 'barfoo.com';
        clientProd.identify({key: key}, null, function() {
          switched = !switched;
          console.log('switched to', key);
          render('prod', 'another-client-side-flag', false);
        });
      });
      document.querySelector('.prod-change-hash').addEventListener('click', function(event) {
        location.hash = new Date().getTime().toString();
      });
      document.querySelector('.prod-change-url').addEventListener('click', function(event) {
        history.pushState(null, null, new Date().getTime().toString());
      });

      //test
      clientTest.on('ready', function() {
        console.log('client is ready');
        main('test');
      });
      
      clientTest.on('change', function(settings) {
        console.log('flags changed:', settings);
        main('test');
      });
      
      clientTest.on('change:another-client-side-flag', function(value, previous) {
        console.log('another-client-side-flag changed:', value, '(' + previous + ')');
        main('test');
      });
      
      document.querySelector('.test-switch-user').addEventListener('click', function(event) {
        var key = switched ? 'foobar.com' : 'barfoo.com';
        clientTest.identify({key: key}, null, function() {
          switched = !switched;
          console.log('switched to', key);
          render('test', 'another-client-side-flag', false);
        });
      });
      document.querySelector('.test-change-hash').addEventListener('click', function(event) {
        location.hash = new Date().getTime().toString();
      });
      document.querySelector('.test-change-url').addEventListener('click', function(event) {
        history.pushState(null, null, new Date().getTime().toString());
      });
    </script>
  </body>
</html>