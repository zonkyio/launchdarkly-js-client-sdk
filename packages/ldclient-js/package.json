{
  "name": "ldclient-js",
  "version": "2.10.2",
  "description": "LaunchDarkly SDK for JavaScript",
  "author": "LaunchDarkly <team@launchdarkly.com>",
  "license": "Apache-2.0",
  "keywords": [
    "launchdarkly",
    "analytics",
    "client"
  ],
  "files": [
    "ldclient.cjs.js",
    "ldclient.cjs.js.map",
    "ldclient.es.js",
    "ldclient.es.js.map",
    "ldclient.min.js",
    "ldclient.min.js.map",
    "typings.d.ts"
  ],
  "types": "./typings.d.ts",
  "main": "dist/ldclient.cjs.js",
  "module": "dist/ldclient.es.js",
  "scripts": {
    "lint": "eslint --format 'node_modules/eslint-formatter-pretty' --ignore-path ../../.eslintignore",
    "lint:all": "eslint --format 'node_modules/eslint-formatter-pretty' --ignore-path ../../.eslintignore src",
    "format": "npm run format:md && npm run format:js",
    "format:md": "prettier --parser markdown --ignore-path ../../.prettierignore --write '*.md'",
    "format:js": "prettier --ignore-path ../../.prettierignore --write 'src/**/*.js'",
    "format:test": "npm run format:test:md && npm run format:test:js",
    "format:test:md": "prettier --parser markdown --ignore-path ../../.prettierignore --list-different '*.md'",
    "format:test:js": "prettier --ignore-path ../../.prettierignore --list-different 'src/**/*.js'",
    "build": "cross-env NODE_ENV=development rollup -c rollup.config.js",
    "build:min": "cross-env NODE_ENV=production rollup -c rollup.config.js",
    "test": "cross-env NODE_ENV=test jest",
    "test:junit": "cross-env NODE_ENV=test jest --testResultsProcessor jest-junit",
    "check-typescript": "node_modules/typescript/bin/tsc",
    "clean": "rimraf dist/**",
    "prepublishOnly": "npm run build:min"
  },
  "devDependencies": {
    "babel-core": "6.26.0",
    "babel-eslint": "8.2.2",
    "babel-jest": "24.3.1",
    "babel-polyfill": "6.26.0",
    "chai": "4.1.2",
    "cross-env": "5.1.4",
    "eslint": "4.19.1",
    "eslint-config-prettier": "2.9.0",
    "eslint-config-xo": "0.20.1",
    "eslint-formatter-pretty": "1.3.0",
    "eslint-plugin-babel": "5.0.0",
    "eslint-plugin-prettier": "2.6.0",
    "jest": "^23.6",
    "jest-junit": "3.6.0",
    "jest-localstorage-mock": "2.2.0",
    "jsdom": "11.11.0",
    "prettier": "1.11.1",
    "readline-sync": "1.4.9",
    "rimraf": "2.6.2",
    "rollup": "0.57.1",
    "semver": "5.5.0",
    "semver-compare": "1.0.0",
    "sinon": "7.2.7",
    "typescript": "3.0.1"
  },
  "dependencies": {
    "escape-string-regexp": "1.0.5",
    "ldclient-js-common": "^2.10.2"
  },
  "repository": {
    "type": "git",
    "url": "git://github.com/launchdarkly/js-client.git"
  }
}
