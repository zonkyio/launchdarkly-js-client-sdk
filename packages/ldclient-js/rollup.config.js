const common = require('../../rollup.common.config');
const pkg = require('./package.json');

const config = Object.assign({}, common, {
  input: 'src/index.js',
  output: [
    {
      plugins: common.plugins,
      name: 'LDClient',
      file: process.env.NODE_ENV === 'production' ? './dist/ldclient.min.js' : './dist/ldclient.js',
      format: 'umd',
      sourcemap: true,
    },
    { file: pkg.main, format: 'cjs', sourcemap: true },
    { file: pkg.module, format: 'es', sourcemap: true },
  ],
});

module.exports = config;
