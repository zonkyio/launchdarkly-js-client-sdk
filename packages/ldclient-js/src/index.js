import * as common from 'ldclient-js-common';
import browserPlatform from './browserPlatform';
import GoalManager from './GoalManager';

const goalsEvent = 'goalsReady';
const extraDefaults = {
  fetchGoals: true,
};

// Pass our platform object to the common code to create the browser version of the client
export function initialize(env, user, options = {}) {
  const platform = browserPlatform();
  const clientVars = common.initialize(env, user, options, platform, extraDefaults);

  const client = clientVars.client;
  const validatedOptions = clientVars.options;
  const emitter = clientVars.emitter;

  const goalsPromise = new Promise(resolve => {
    const onGoals = emitter.on(goalsEvent, () => {
      emitter.off(goalsEvent, onGoals);
      resolve();
    });
  });
  client.waitUntilGoalsReady = () => goalsPromise;

  if (validatedOptions.fetchGoals) {
    const goalManager = GoalManager(clientVars, () => emitter.emit(goalsEvent));
    platform.customEventFilter = goalManager.goalKeyExists;
  } else {
    emitter.emit(goalsEvent);
  }

  if (document.readyState !== 'complete') {
    window.addEventListener('load', clientVars.start);
  } else {
    clientVars.start();
  }
  window.addEventListener('beforeunload', () => {
    platform.pageIsClosing = true;
    client.close();
  });

  return client;
}

export const createConsoleLogger = common.createConsoleLogger;

export const version = common.version;

function deprecatedInitialize(env, user, options = {}) {
  console && console.warn && console.warn(common.messages.deprecated('default export', 'named LDClient export'));
  return initialize(env, user, options);
}

export default { initialize: deprecatedInitialize, version };
