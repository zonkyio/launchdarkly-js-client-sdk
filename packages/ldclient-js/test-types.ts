
// This file exists only so that we can run the TypeScript compiler in the CI build
// to validate our typings.d.ts file.

import * as ld from 'ldclient-js';

var ver: string = ld.version;

var emptyOptions: ld.LDOptions = {};
var logger: ld.LDLogger = ld.createConsoleLogger("info");
var allOptions: ld.LDOptions = {
  bootstrap: { },
  hash: '',
  baseUrl: '',
  eventsUrl: '',
  streamUrl: '',
  streaming: true,
  useReport: true,
  sendLDHeaders: true,
  evaluationReasons: true,
  fetchGoals: true,
  sendEvents: true,
  allAttributesPrivate: true,
  privateAttributeNames: [ 'x' ],
  allowFrequentDuplicateEvents: true,
  sendEventsOnlyForVariation: true,
  flushInterval: 1,
  samplingInterval: 1,
  streamReconnectDelay: 1,
  logger: logger
};
var userWithKeyOnly: ld.LDUser = { key: 'user' };
var user: ld.LDUser = {
  key: 'user',
  name: 'name',
  firstName: 'first',
  lastName: 'last',
  email: 'test@example.com',
  avatar: 'http://avatar.url',
  ip: '1.1.1.1',
  country: 'us',
  anonymous: true,
  custom: {
    'a': 's',
    'b': true,
    'c': 3,
    'd': [ 'x', 'y' ],
    'e': [ true, false ],
    'f': [ 1, 2 ]
  },
  privateAttributeNames: [ 'name', 'email' ]
};
var client: ld.LDClient = ld.initialize('env', user, allOptions);

var boolFlagValue: ld.LDFlagValue = client.variation('key', false);
var numberFlagValue: ld.LDFlagValue = client.variation('key', 2);
var stringFlagValue: ld.LDFlagValue = client.variation('key', 'default');

var detail: ld.LDEvaluationDetail = client.variationDetail('key', 'default');
var detailValue: ld.LDFlagValue = detail.value;
var detailIndex: number | undefined = detail.variationIndex;
var detailReason: ld.LDEvaluationReason = detail.reason;

var flagSet: ld.LDFlagSet = client.allFlags();
var flagSetValue: ld.LDFlagValue = flagSet['key'];
