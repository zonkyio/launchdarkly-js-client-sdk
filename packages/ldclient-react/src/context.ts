import { createContext } from 'react';
import { LDClient, LDFlagSet } from 'ldclient-js';

/**
 * The LaunchDarkly context stored in the Provider state and passed to consumers.
 */
interface LDContext {
  /**
   * Contains all flags from LaunchDarkly. This object will always exist but will be empty {} initially
   * until flags are fetched from the LaunchDarkly servers.
   */
  flags: LDFlagSet;

  /**
   * The LaunchDarkly client's instance interface. This will be be undefined initially until
   * initialization is complete.
   *
   * @see http://docs.launchdarkly.com/docs/js-sdk-reference
   */
  ldClient?: LDClient;
}

const { Provider, Consumer } = createContext<LDContext>({ flags: {}, ldClient: undefined });
export { Provider, Consumer, LDContext };
