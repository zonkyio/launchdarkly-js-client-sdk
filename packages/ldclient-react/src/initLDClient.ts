import { initialize as ldClientInitialize, LDClient, LDFlagSet, LDOptions, LDUser } from 'ldclient-js';
import { v4 as uuid } from 'uuid';
import { camelCaseKeys } from './utils';

interface AllFlagsLDClient {
  flags: LDFlagSet;
  ldClient: LDClient;
}

const initLDClient = async (
  clientSideID: string,
  user: LDUser = { key: uuid() },
  options?: LDOptions,
  targetFlags?: LDFlagSet,
): Promise<AllFlagsLDClient> => {
  const ldClient = ldClientInitialize(clientSideID, user, options);

  return new Promise<AllFlagsLDClient>(resolve => {
    ldClient.on('ready', () => {
      let rawFlags: LDFlagSet = {};

      if (targetFlags) {
        for (const flag in targetFlags) {
          rawFlags[flag] = ldClient.variation(flag, targetFlags[flag]);
        }
      } else {
        rawFlags = ldClient.allFlags();
      }

      resolve({ flags: camelCaseKeys(rawFlags), ldClient });
    });
  });
};

export default initLDClient;
