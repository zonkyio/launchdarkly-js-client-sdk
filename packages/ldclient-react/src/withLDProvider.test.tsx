jest.mock('./initLDClient', () => jest.fn());
jest.mock('./context', () => ({ Provider: 'Provider' }));

import * as React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import { LDFlagChangeset, LDFlagSet, LDOptions, LDUser } from 'ldclient-js';
import initLDClient from './initLDClient';
import withLDProvider, { EnhancedComponent } from './withLDProvider';
import { LDContext as HocState } from './context';

const clientSideID = 'deadbeef';
const App = () => <div>My App</div>;
const mockInitLDClient = initLDClient as jest.Mock;
const mockFlags = { testFlag: true, anotherTestFlag: true };
const mockLDClient = {
  on: jest.fn((e: string, cb: () => void) => {
    cb();
  }),
};

describe('withLDProvider', () => {
  beforeEach(() => {
    mockInitLDClient.mockImplementation(() => ({
      flags: mockFlags,
      ldClient: mockLDClient,
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('render app', async () => {
    const LaunchDarklyApp = withLDProvider({ clientSideID })(App);
    const component = create(<LaunchDarklyApp />);
    expect(component).toMatchSnapshot();
  });

  test('ld client is initialised correctly', async () => {
    const user: LDUser = { key: 'yus', name: 'yus ng' };
    const options: LDOptions = { bootstrap: {} };
    const LaunchDarklyApp = withLDProvider({ clientSideID, user, options })(App);
    const instance = create(<LaunchDarklyApp />).root.instance as EnhancedComponent;

    await instance.componentDidMount();
    expect(mockInitLDClient).toHaveBeenCalledWith(clientSideID, user, options, undefined);
  });

  test('ld client is bootstrapped correctly', async () => {
    const user: LDUser = { key: 'yus', name: 'yus ng' };
    const options: LDOptions = {
      bootstrap: {
        'test-flag': true,
        'another-test-flag': false,
        $flagsState: {
          'test-flag': { version: 125, variation: 0, trackEvents: true },
          'another-test-flag': { version: 18, variation: 1 },
        },
        $valid: true,
      },
    };
    const LaunchDarklyApp = withLDProvider({ clientSideID, user, options })(App);
    const component = shallow(<LaunchDarklyApp />, { disableLifecycleMethods: true });
    const initialState = component.state() as HocState;

    expect(mockInitLDClient).not.toHaveBeenCalled();
    expect(initialState.flags).toEqual({ testFlag: true, anotherTestFlag: false });
  });

  test('state.flags should be initialised to empty when bootstrapping from localStorage', async () => {
    const user: LDUser = { key: 'yus', name: 'yus ng' };
    const options: LDOptions = {
      bootstrap: 'localStorage',
    };
    const LaunchDarklyApp = withLDProvider({ clientSideID, user, options })(App);
    const component = shallow(<LaunchDarklyApp />, { disableLifecycleMethods: true });
    const initialState = component.state() as HocState;

    expect(mockInitLDClient).not.toHaveBeenCalled();
    expect(initialState.flags).toEqual({});
  });

  test('ld client is initialised correctly with target flags', async () => {
    mockInitLDClient.mockImplementation(() => ({
      flags: { devTestFlag: true, launchDoggly: true },
      ldClient: mockLDClient,
    }));
    const user: LDUser = { key: 'yus', name: 'yus ng' };
    const options: LDOptions = { bootstrap: {} };
    const flags = { 'dev-test-flag': false, 'launch-doggly': false };
    const LaunchDarklyApp = withLDProvider({ clientSideID, user, options, flags })(App);
    const instance = create(<LaunchDarklyApp />).root.instance as EnhancedComponent;
    instance.setState = jest.fn();

    await instance.componentDidMount();

    expect(mockInitLDClient).toHaveBeenCalledWith(clientSideID, user, options, flags);
    expect(instance.setState).toHaveBeenCalledWith({
      flags: { devTestFlag: true, launchDoggly: true },
      ldClient: mockLDClient,
    });
  });

  test('flags and ldClient are saved in state on mount', async () => {
    const LaunchDarklyApp = withLDProvider({ clientSideID })(App);
    const instance = create(<LaunchDarklyApp />).root.instance as EnhancedComponent;
    instance.setState = jest.fn();

    await instance.componentDidMount();
    expect(instance.setState).toHaveBeenCalledWith({ flags: mockFlags, ldClient: mockLDClient });
  });

  test('subscribeToChanges is called on mount', async () => {
    const LaunchDarklyApp = withLDProvider({ clientSideID })(App);
    const instance = create(<LaunchDarklyApp />).root.instance as EnhancedComponent;
    instance.subscribeToChanges = jest.fn();

    await instance.componentDidMount();
    expect(instance.subscribeToChanges).toHaveBeenCalled();
  });

  test('subscribe to changes', async () => {
    mockLDClient.on.mockImplementation((e: string, cb: (c: LDFlagChangeset) => void) => {
      cb({ 'test-flag': { current: false, previous: true } });
    });
    const LaunchDarklyApp = withLDProvider({ clientSideID })(App);
    const instance = create(<LaunchDarklyApp />).root.instance as EnhancedComponent;
    const mockSetState = jest.spyOn(instance, 'setState');

    await instance.componentDidMount();
    const callback = mockSetState.mock.calls[1][0] as (flags: LDFlagSet) => LDFlagSet;
    const newState = callback({ flags: mockFlags });

    expect(mockLDClient.on).toHaveBeenCalledWith('change', expect.any(Function));
    expect(newState).toEqual({ flags: { anotherTestFlag: true, testFlag: false } });
  });
});
