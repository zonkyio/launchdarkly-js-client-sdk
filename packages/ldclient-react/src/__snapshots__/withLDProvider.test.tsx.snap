// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLDProvider render app 1`] = `
<Provider
  value={
    Object {
      "flags": Object {},
      "ldClient": undefined,
    }
  }
>
  <div>
    My App
  </div>
</Provider>
`;
